### PR TITLE
Logging Code Information

### PR DESCRIPTION
**Pull request by didvi**
Updated logging of exceptions to include:

"codeToLinenos" a list of tuples containing the line of code that caused the error and the line number this code was on. this is saved for the first two frames of the traceback.

"traceSummary" the traceback for the error formatted as an array, where each item is a line of the traceback. This can easily be printed by iterating over this array and printing out each element. This includes the entire traceback, unlike the "codeToLinenos"

See errorLog.csv for an example of the logging.

### CODE REVIEW

File: d8error.py [hunk 1]
```diff
@@ -32,29 +34,36 @@ def __init__(self, etype, value):
...
-            """saves errors to errorLog.csv"""
-            fieldnames = ['index', 'errorType', 'errorMSG', 'feedbackRating', 'feedbackMSG', "Time"]
-            writer = csv.DictWriter(file, fieldnames=fieldnames)
+        # this generates a semi-readable summary of the traceback, which includes some information about the python code that caused the error
+        summary = traceback.format_exc().splitlines()
+        
+        # iterate through traceback object to extract linenumber and bytecode of the first two frames
+        curr_tb = tb.tb_next # skip the first frame which is the jupyter notebook frame
+        bytecodesToLinenos = []
+        while curr_tb and len(bytecodesToLinenos) < 2:
```
Comment (micahtyong) [inline d8error.py]: Without the `< 2` constraint, does this loop go further into the stack trace?

File: d8error.py [hunk 2]
```diff
@@ -32,29 +34,36 @@ def __init__(self, etype, value):
...
+            fieldnames = ['index', 'errorType', 'errorMSG', 'feedbackRating', 'feedbackMSG', 'time', 'bytecodesToLinenos', 'traceSummary']
+            writer = csv.DictWriter(f, fieldnames=fieldnames)
             writer.writerow({"index": self.eindex,
                             "errorType": self.errorname,
                             "errorMSG": str(self.value),
                             "feedbackRating": self.feedbackRating,
                             "feedbackMSG": self.feedbackMSG,
-                            "Time": str(datetime.datetime.now())})
-
-        if not os.path.isfile("errorLog.csv"):
```
Comment (micahtyong) [inline d8error.py:46]: Was this code meant to be deleted?

File: errorLog.csv [hunk 3]
```diff
@@ -0,0 +1,14 @@
+1,TypeError,'int' object is not callable,0,,2021-11-07 15:55:34.668983,"[(b'd\x00d\x01\x83\x01F\x00d\x02S\x00', 1)]","['  File ""/usr/local/lib/python3.8/site-packages/IPython/core/interactiveshell.py"", line 3417, in run_code\n    exec(code_obj, self.user_global_ns, self.user_ns)\n', '  File ""<ipython-input-5-9b43eec38a3d>"", line 1, in <module>\n    3(2)\n']"
```
Comment (micahtyong) [inline errorLog.csv]: If we don't get the Python decompyle to work, we should probably just remove the bytecode column. cc: @kenny-chi

File: d8error.py (hunk 1 above)
Comment (didvi) [reply d8error.py]: yes but Raymond mentioned that he only wanted the first two

File: d8error.py (hunk 2 above)
Comment (didvi) [reply d8error.py:46]: yes i just cleaned it up to make it compatible with my changes

**Review by micahtyong — APPROVED** [review]

**Pull request merged**